File: leetcode/lc-us/cpp/1019.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    vector<int> nextLargerNodes(ListNode *head) {
        vector<int> v;
        while (head) {
            v.push_back(head->val);
            head = head->next;
        }
        auto n = v.size();
        vector<int> res(n);
        stack<int> stk;
        for (int i = 0; i < n; i++) {
            while (!stk.empty() && v[stk.top()] < v[i]) {
                if (v[stk.top()] < v[i]) res[stk.top()] = v[i];
                stk.pop();
            }
            stk.push(i);
        }
        return res;
    }
};

File: leetcode/lc-us/rust/1019.rs
impl Solution {
    pub fn next_larger_nodes(head: Option<Box<ListNode>>) -> Vec<i32> {
        let mut v: Vec<i32> = Vec::new();
        let mut head = head.as_ref();
        while head != None {
            v.push(head.unwrap().val);
            head = head.unwrap().next.as_ref();
        }
        let n = v.len();
        let mut res: Vec<i32> = vec![0; n];
        let mut stack: Vec<usize> = Vec::new();
        for i in 0..n {
            while !stack.is_empty() && v[stack[stack.len() - 1]] < v[i] {
                res[stack[stack.len() - 1] as usize] = v[i];
                stack.pop();
            }
            stack.push(i);
        }
        res
    }
}